File: c_programming/str/n45_find_the_index_of_the_first_occurrence_in_a_string_28.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include "utils.h"

static size_t is_compare(const char *in_str, const char *pat, bool *result)
{
    size_t i = 0;
    size_t pat_len = strlen(pat);
    char mat_c = pat[0];
    size_t mat_index = 0;

    *result = true;
    for (i = 0; i < pat_len; i ++) {
        if (in_str[i] == mat_c && mat_index == 0) {
            mat_index = i;
            if (*result == false) {
                goto finish;
            }
        }
        if (*result == false || in_str[i] != pat[i]) {
            *result = false;
            if (mat_index == 0) {
                continue;
            } else {
                goto finish;
            }
        }
    }

finish:
    return mat_index;
}

static int32_t find_sub_str(const char *in_str, const char *pat, int32_t *index)
{
    int32_t ret = 0;
    size_t in_len = 0;
    size_t pat_len = 0;
    size_t i = 0;
    size_t mat_index = 0;
    bool res = false;

    UTILS_CHECK_PTR(in_str);
    UTILS_CHECK_PTR(pat);
    UTILS_CHECK_PTR(index);
    UTILS_CHECK_PTR(pat_len = strlen(pat));
    UTILS_CHECK_PTR(in_len = strlen(in_str));

    while (i < in_len) {
        mat_index = is_compare(in_str + i, pat, &res);
        LOG("find compare %s, the result is %d, the next index is +%d\n", in_str + i, res, mat_index);
        if (res == true) {
            *index = (int32_t)(i + mat_index);
            goto finish;
        }
        i += (mat_index == 0) ? 1 : mat_index;

    }

    *index = -1;

finish:
    return ret;
}

int32_t main(void)
{
    int32_t ret = 0;
    int32_t index = 0;

    ret = find_sub_str("adesaddadbutsad", "sad", &index);
    UTILS_CHECK_RET(ret);

    LOG("the output is %d\n", index);

finish:
    return ret;
}

File: rust_programming/str/src/n45_find_the_index_of_the_first_occurrence_in_a_string_28.rs


fn is_compare(in_str:&str, pat:&str, res:&mut bool) -> i32
{
    let mut in_chars = in_str.chars();
    let mut pat_chars = pat.chars();
    let mat_c:char = pat_chars.nth(0).unwrap();
    let mut i:i32 = 0;
    let mut mat_index:i32 = 0;

    *res = true;
    for e in pat_chars {
        if e == mat_c && 0 == mat_index {
            mat_index = i;
            if false == *res {
                return mat_index;
            }
        }
        if false == *res || e != in_chars.nth(i as usize).unwrap() {
            *res = false;
            if 0 == mat_index {
                continue;
            } else {
                return mat_index;
            }
        }
        i += 1;
    }

    return mat_index;
}

fn find_sub_str(in_str:&String, pat:&String) -> i32
{
    let mut mat_index:i32;
    let mut i:usize = 0;
    let in_len:usize = in_str.len();
    let mut res:bool = true;

    while i < in_len {
        mat_index = is_compare(&in_str[i..], &pat[0..], &mut res);
        if true == res {
            return i as i32 + mat_index;
        }
        if 0 == mat_index {
            i += 1;
        } else {
            i += mat_index as usize;
        }
    }

    mat_index = -1;

    return mat_index;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    pub fn test_most_common_word() {
        let ret:i32 = find_sub_str(&"adesaddadbutsad".to_string(), &"sad".to_string());
        assert_eq!(ret, 4);
    }
}